ext/typeexpr/defaults_test.go

```diff
@@ -649,6 +649,105 @@ func TestDefaults_Apply(t *testing.T) {
 			value: cty.EmptyObjectVal,
 			want:  cty.EmptyObjectVal,
 		},
+		"tuples retain custom values and dynamic types": {
+			defaults: &Defaults{
+				Type: cty.List(cty.ObjectWithOptionalAttrs(map[string]cty.Type{
+					"name":   cty.String,
+					"taints": cty.List(cty.Map(cty.DynamicPseudoType)),
+				}, []string{"name", "taints"})),
+				Children: map[string]*Defaults{
+					"": {
+						Type: cty.ObjectWithOptionalAttrs(map[string]cty.Type{
+							"name":   cty.String,
+							"taints": cty.List(cty.Map(cty.DynamicPseudoType)),
+						}, []string{"name", "taints"}),
+						DefaultValues: map[string]cty.Value{
+							"name":   cty.StringVal("default"),
+							"taints": cty.ListValEmpty(cty.Map(cty.DynamicPseudoType)),
+						},
+					},
+				},
+			},
+			value: cty.TupleVal([]cty.Value{
```

Comment:
Since `Defaults.Apply` requires the caller to convert to the target type before passing in a value, I think this test case is effectively "impossible". If we were to add the check I mentioned in my previous comment then this would panic, but I think that's okay because the subsequent test `"lists merge dynamic types with concrete types"` is covering the valid equivalent where the caller has already asked `convert.Convert` to turn the tuple value into a list value.

Reply:
I've modified the docs slightly to reflect that conversion should happen after, which I think means this is a valid use case again.